## **Reading Food Hygiene Data Analysis**

### Objectives

* Analyse the UK Food Standards Agency dataset, and clean in preparation for data analysis.

### Planned Steps
1. Download and load raw data.
2. Filter data to only Reading Local Authority
3. Identify columns needed for analysis - Create new Reading dataset
4. Initial inspection of data
5. Clean - Remove incomplete or irrelevant data
6. Validate clean dataset
7. Export cleaned dataset for analysis, visualisation and conclusion

### Inputs (Data Source)
* **Data Source:** [UK Food Standards Agency - Open Data CSV](https://safhrsprodstorage.blob.core.windows.net/opendatafileblobstorage/FHRS_All_en-GB.csv)

### Outputs
* (3) - Filtered Reading Local Authority dataset
* (6) - Cleaned Reading Local Authority dataset for analysis

---

### Setup
Install and Import required libraries

In [176]:
%pip install pandas numpy

import numpy as np
import pandas as pd

print("Libraries imported successfully!")

Note: you may need to restart the kernel to use updated packages.
Libraries imported successfully!



[notice] A new release of pip is available: 24.3.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


---
### 1. Download and load raw data.
Raw data downloaded from the Food Standards Agency (link above).

**File Location:** Saved locally in the `data` directory as `FHRS_All_en-GB.csv`

In [177]:
file_path = "data/FHRS_All_en-GB.csv"  # Set the file path to the CSV file
print(f"Reading data from {file_path}...")  # Print a message indicating the file being read
df_raw = pd.read_csv(file_path, encoding="utf-8-sig", low_memory=False)  # Read the CSV file into a DataFrame
print(f"Data read successfully. Shape of the DataFrame: {df_raw.shape}")  # Print the shape of the DataFrame

Reading data from data/FHRS_All_en-GB.csv...
Data read successfully. Shape of the DataFrame: (611521, 23)


---
### 2. Filter data to only Reading Local Authority
- Filter data to Reading

In [178]:
# Identify the name of the columns to target local authority
df_raw.columns.tolist()  

['AddressLine1',
 'AddressLine2',
 'AddressLine3',
 'AddressLine4',
 'BusinessTypeID',
 'FHRSID',
 'BusinessName',
 'BusinessType',
 'ConfidenceInManagement',
 'Hygiene',
 'Latitude',
 'LocalAuthorityBusinessID',
 'LocalAuthorityCode',
 'LocalAuthorityName',
 'Longitude',
 'NewRatingPending',
 'PostCode',
 'RatingDate',
 'RatingKey',
 'RatingValue',
 'RightToReply',
 'SchemeType',
 'Structural']

In [179]:
# Identify the actual name of the local authority, and sort alphabetically

unique_authorities = df_raw['LocalAuthorityName'].unique().tolist()  # List of unique local authorities
unique_authorities.sort()  # Sort the list alphabetically 
unique_authorities

['Aberdeen City',
 'Aberdeenshire',
 'Adur',
 'Amber Valley',
 'Anglesey',
 'Angus',
 'Antrim and Newtownabbey',
 'Ards and North Down',
 'Argyll and Bute',
 'Armagh City, Banbridge and Craigavon',
 'Arun',
 'Ashfield',
 'Ashford',
 'Babergh',
 'Barking and Dagenham',
 'Barnet',
 'Barnsley',
 'Basildon',
 'Basingstoke and Deane',
 'Bassetlaw',
 'Bath and North East Somerset',
 'Bedford',
 'Belfast City',
 'Bexley',
 'Birmingham',
 'Blaby',
 'Blackburn',
 'Blackpool',
 'Blaenau Gwent',
 'Bolsover',
 'Bolton',
 'Boston',
 'Bournemouth, Christchurch and Poole',
 'Bracknell Forest',
 'Bradford',
 'Braintree',
 'Breckland',
 'Brent',
 'Brentwood',
 'Bridgend',
 'Brighton and Hove',
 'Bristol',
 'Broadland',
 'Bromley',
 'Bromsgrove',
 'Broxbourne',
 'Broxtowe',
 'Buckinghamshire',
 'Burnley',
 'Bury',
 'Caerphilly',
 'Calderdale',
 'Cambridge City',
 'Camden',
 'Cannock Chase',
 'Canterbury City',
 'Cardiff',
 'Carmarthenshire',
 'Castle Point',
 'Causeway Coast and Glens',
 'Central Bedfor

---
 ### 3. Identify columns needed for analysis
We already know the names of the columns from Step 2.

The columns to keep:
| Column Name | What it is | Rationale |
| :--- | :--- | :--- |
| LocalAuthorityBusinessID | Unique reference code given to the establishment | Serves as a unique identifier for the business
| BusinessName | Trading name of the business | Essential for labelling charts and specific lookups
| BusinessType | The category of the business (e.g., Restaurant/Cafe/Retailer) | Crucial for comparative analysis across sectors
| RatingValue | The hygiene score awarded to the business | Primary target metric for distributions and profiling
| RatingDate | The date when the inspection took place | Enables analysis of inspection trends
| AddressLine1 | The first line of the address for the business | Adds granular location context for individual venues 
| Postcode | The postcode for the business | Useful for neighbourhood grouping and mapping
| LocalAuthorityName | The name of the Local Authority (Reading) | Verification anchor confirming filter accuracy

Save new dataset

In [180]:
reading_data = (df_raw[df_raw['LocalAuthorityName'] == 'Reading']  # Filter the DataFrame for Reading local authority
[['LocalAuthorityBusinessID', 'BusinessName', 'BusinessType', 'RatingValue', 'RatingDate', 'AddressLine1', 'PostCode', 'LocalAuthorityName']].copy())  # Select relevant columns

reading_data.to_csv('data/reading_food_hygiene_data.csv', index=False)  # Save the filtered DataFrame to a new CSV file

display(reading_data.head())  # Display the first few rows of the filtered DataFrame

,LocalAuthorityBusinessID,BusinessName,BusinessType,RatingValue,RatingDate,AddressLine1,PostCode,LocalAuthorityName
413879,137934,2 Chefs Catering Ltd,Other catering premises,5,2024-11-24,NaN,NaN,Reading
413880,80333,3 Love Food,Other catering premises,5,2026-04-17,NaN,NaN,Reading
413881,142094,3B's Cafe,Other catering premises,5,2026-03-17,735 Oxford Road,RG30 1JA,Reading
413882,127516,3Cs Cafe at St Johns Church,Restaurant/Cafe/Canteen,5,2025-09-23,Gosbrook Road,RG4 5AN,Reading
413883,72396,6Pack Supplements,Distributors/Transporters,Exempt,NaN,443 Oxford Road,RG30 1HD,Reading


---
### Dataset Optimisation
By filtering the nationwide UK Food Standards Agency raw dataset **(137MB)**, to our target area of **Reading** and selecting only the 8 target columns, the processed dataset has been streamlined to just **143KB**.

**Why this matters:** This drastically speeds up execution times for visualisation tools, while keeping our project lightweight. 

**File Location:** Saved locally in the `data` directory as `reading_food_hygiene_data.csv`

---
### 4. Inspect and clean data ready for analysis
- Load `reading_food_hygiene_data.csv`
- Check the initial shape of the dataset
- Check the columns, data type and non-null counts
- Check for missing values across all columns
- Check the percentage of missing data for each column

In [181]:
file_path = "data/reading_food_hygiene_data.csv"  # Set the file path to the new CSV file
df = pd.read_csv(file_path)  # Read the CSV file into a DataFrame

print(f'Shape:\n{df.shape}')  # Print the initial shape of the dataframe

Shape:
(1508, 8)


In [182]:
df.isnull().sum()  # Print the number of null values in each column

LocalAuthorityBusinessID      0
BusinessName                  0
BusinessType                  0
RatingValue                   0
RatingDate                  188
AddressLine1                158
PostCode                    187
LocalAuthorityName            0
dtype: int64

In [183]:
df.isnull().sum() / len(df) * 100  # Print the percentage of null values in each column

LocalAuthorityBusinessID     0.000000
BusinessName                 0.000000
BusinessType                 0.000000
RatingValue                  0.000000
RatingDate                  12.466844
AddressLine1                10.477454
PostCode                    12.400531
LocalAuthorityName           0.000000
dtype: float64

In [184]:
df.describe(include='all') # Print the statistics for the dataset

,LocalAuthorityBusinessID,BusinessName,BusinessType,RatingValue,RatingDate,AddressLine1,PostCode,LocalAuthorityName
count,1508,1508,1508,1508,1320,1350,1321,1508
unique,1508,1422,13,8,553,1121,590,1
top,137934,Tesco,Restaurant/Cafe/Canteen,5,2026-03-11,Bridge Street,RG1 2AG,Reading
freq,1,10,334,853,10,37,28,1508


---
### Missing Data
By the above, we can see that we are missing values in the following columns:
- 188 in RatingDate
- 158 in AddressLine1
- 187 in PostCode

We now need to look at these columns to check what the missing data is, and make a decision on what to do with them

In [186]:
# Filter to the rows where any columns have null values
missing_rows = df[df['RatingDate'].isnull() | df['AddressLine1'].isnull() | df['PostCode'].isnull()]
missing_rows


,LocalAuthorityBusinessID,BusinessName,BusinessType,RatingValue,RatingDate,AddressLine1,PostCode,LocalAuthorityName
0,137934,2 Chefs Catering Ltd,Other catering premises,5,2024-11-24,NaN,NaN,Reading
1,80333,3 Love Food,Other catering premises,5,2026-04-17,NaN,NaN,Reading
4,72396,6Pack Supplements,Distributors/Transporters,Exempt,NaN,443 Oxford Road,RG30 1HD,Reading
12,99401,A & G Retail,Retailers - other,5,2024-06-03,NaN,NaN,Reading
13,144943,A J Bakes & More,Other catering premises,4,2025-05-08,NaN,NaN,Reading
...,...,...,...,...,...,...,...,...
1473,06916,Woodley Grill,Mobile caterer,AwaitingInspection,NaN,NaN,NaN,Reading
1480,06793,Yan's Kitchen,Other catering premises,AwaitingInspection,NaN,NaN,NaN,Reading
1482,140746,Yattendon Honey,Other catering premises,5,2025-09-09,NaN,NaN,Reading
1501,06962,Zeus Greek Gyros,Restaurant/Cafe/Canteen,AwaitingInspection,NaN,2 Bridge Street,RG1 2LU,Reading


initial thoughts
The busineesses that are missing dates appear to be due to awaiting inspection or exempt - in that regard as they serve no value to answering our questions - consider removal